fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanAccrualActivityProcessingServiceImpl.java

```diff
@@ -46,58 +51,93 @@
+        List<LoanRepaymentScheduleInstallment> installments = loan
+                .getRepaymentScheduleInstallments(i -> !i.isDownPayment() && !DateUtils.isBefore(currentDate, i.getDueDate()));
+        for (LoanRepaymentScheduleInstallment installment : installments) {
+            LocalDate dueDate = installment.getDueDate();
+            // check if there is any not-replayed-accrual-activity related to business date
+            ArrayList<LoanTransaction> existingActivities = new ArrayList<>(
+                    loan.getLoanTransactions(t -> t.isNotReversed() && t.isAccrualActivity() && t.getTransactionDate().isEqual(dueDate)));
+            if (existingActivities.isEmpty()) {
+                makeAccrualActivityTransaction(loan, installment, dueDate);
+            } else {
+                LoanTransaction existingActivity = existingActivities.get(0);
+                reverseReplayAccrualActivityTransaction(loan, existingActivity, installment, dueDate);
+                existingActivities.remove(existingActivity);
+                existingActivities.forEach(this::reverseAccrualActivityTransaction);
             }
         }
-        return loan;
     }
 
     @Override
     @Transactional
-    public void processAccrualActivityForLoanClosure(Loan loan) {
+    public void processAccrualActivityForLoanClosure(@NotNull Loan loan) {
+        if (!loan.getLoanProductRelatedDetail().isEnableAccrualActivityPosting()) {
+            return;
+        }
         LocalDate date = loan.isOverPaid() ? loan.getOverpaidOnDate() : loan.getClosedOnDate();
         List<LoanTransaction> accrualActivityTransaction = loan.getLoanTransactions().stream().filter(LoanTransaction::isNotReversed)
                 .filter(LoanTransaction::isAccrualActivity).filter(loanTransaction -> loanTransaction.getDateOf().isAfter(date)).toList();
         if (!accrualActivityTransaction.isEmpty()) {
             accrualActivityTransaction.forEach(this::reverseAccrualActivityTransaction);
         }
         LoanTransaction loanTransaction = assembleClosingAccrualActivityTransaction(loan, date);
-        if (loanTransaction.getAmount().compareTo(BigDecimal.ZERO) != 0) {
-            loanWritePlatformService.makeAccrualActivityTransaction(loan, loanTransaction);
+        if (loanTransaction != null) {
+            makeAccrualActivityTransaction(loan, loanTransaction);
         }
     }
 
-    private void reverseAccrualActivityTransaction(LoanTransaction loanTransaction) {
-        loanChargeValidator.validateRepaymentTypeTransactionNotBeforeAChargeRefund(loanTransaction.getLoan(), loanTransaction, "reversed");
-        loanTransaction.reverse();
-        LoanAdjustTransactionBusinessEvent.Data data = new LoanAdjustTransactionBusinessEvent.Data(loanTransaction);
-        businessEventNotifierService.notifyPostBusinessEvent(new LoanAdjustTransactionBusinessEvent(data));
+    @Override
+    @Transactional
+    public void processAccrualActivityForLoanReopen(@NotNull Loan loan) {
```

Comment:
As far as I can tell this method existed before but now it's impossible to verify what was changed here exactly. Can you please revert the method reordering (same thing for others if it happened)?

if reordering makes sense, feel free to open a separate PR just for that.

Reply:
Common private methods were moved to the utility section, following the call order.

Reply:
I honestly disagree with it and the reason of not reverting the method movement (rewrote most of the methods) doesn't have a ground. Even if you refactored the method bodies, you can revert the method movement - making it possible to review it.

I'll approve it just to get the ball going but next time, please keep in mind to not do unnecessary changes that makes PR reviews hell.